lib/utils/compose.coffee

```diff
@@ -0,0 +1,732 @@
+	_ = require('lodash')
+	querystring = require('querystring')
+	request = require('request')
+	getAsync = Promise.promisify(request.get)
+
+	if not _.isArray(images)
+		images = [ images ]
+
+	query = querystring.stringify
+		service: registry
+		scope: images.map (repo) ->
+			"repository:#{repo}:pull,push"
+
+	getAsync
+		url: "#{tokenAuthEndpoint}/auth/v1/token?#{query}"
+		json: true
+		headers: Authorization: auth
+	.get('body')
+	.get('token')
+	.catchReturn({})
+
+pushAndUpdateServiceImages = (docker, token, images, afterEach) ->
+	chalk = require('chalk')
+	{ DockerProgress } = require('docker-progress')
+	tty = require('./tty')(process.stdout)
+
+	opts = { authconfig: registrytoken: token }
+
+	progress = new DockerProgress(dockerToolbelt: docker)
+	renderer = pushProgressRenderer(tty, chalk.blue('[Push]') + '    ')
+	reporters = progress.aggregateProgress(images.length, renderer)
+
+	Promise.using tty.cursorHidden(), ->
+		Promise.map images, ({ serviceImage, localImage, props }, index) ->
+			Promise.join(
+				localImage.inspect().get('Size')
+				progress.push(localImage.name, reporters[index], opts).finally(renderer.end)
+				(size, digest) ->
+					serviceImage.image_size = size
+					serviceImage.content_hash = digest
+					serviceImage.dockerfile = props.dockerfile
+					serviceImage.project_type = props.projectType
+					serviceImage.start_timestamp = props.startTime if props.startTime
+					serviceImage.end_timestamp = props.endTime if props.endTime
+					serviceImage.push_timestamp = new Date()
+					serviceImage.status = 'success'
+			)
+			.catch (e) ->
+				serviceImage.error_message = '' + e
+				serviceImage.status = 'failed'
+				throw e
+			.finally ->
+				afterEach?(serviceImage, props)
+
+exports.deployProject = (
+	docker, logger,
+	composition, images,
+	appId, userId, auth,
+	apiEndpoint,
+	skipLogUpload
+) ->
+	_ = require('lodash')
+	releaseMod = require('resin-release')
+
+	logger.logInfo('Creating release...')
+
+	createRelease(apiEndpoint, auth, userId, appId, composition)
+	.then ({ client, release, serviceImages }) ->
+		logger.logDebug('Tagging images...')
+		tagServiceImages(docker, images, serviceImages)
+		.tap (images) ->
+			logger.logDebug('Authorizing push...')
+			authorizePush(apiEndpoint, auth, images[0].registry, _.map(images, 'repo'))
+			.then (token) ->
+				logger.logInfo('Pushing images to registry...')
+				pushAndUpdateServiceImages docker, token, images, (serviceImage) ->
+					logger.logDebug("Saving image #{serviceImage.is_stored_at__image_location}")
+					releaseMod.updateImage(client, serviceImage.id, serviceImage)
+			.finally ->
+				logger.logDebug('Untagging images...')
+				Promise.map images, ({ localImage }) ->
+					localImage.remove()
+		.tap (images) ->
+			return if skipLogUpload
+			logger.logInfo('Uploading build logs...')
+
+			Promise.map images, ({ serviceName, serviceImage, logs }) ->
+				releaseMod.updateImage(client, serviceImage.id, build_log: logs)
```

Comment:
No reason, good point, I loosely followed what the previous command did, i.e uploading logs separately in the end.

Reply:
That was mainly due to the implementation having two seperate endpoints, but I think with the setup we're doing now, it makes sense to push it as a single request. Also it's worth noting you'll have to make sure that the payload doesn't exceed 512Kb, otherwise the API will reject it.